In [7]:
import pandas as pd
import os
import re
import smtplib
from email.mime.text import MIMEText
from email.mime.multipart import MIMEMultipart

In [8]:
def parse_logs(filename):
    with open(filename) as f:
        content = f.readlines()        
        
    content = [x.strip() for x in content] 
    results = []
    
    for line in content:    
        try:
            data = {}
            data['passed'] = bool(re.search("passed", line, flags=re.IGNORECASE))
            data['name'] = re.search("isis3_[A-z]+|[A-z]+_app_[A-z]+", line, flags=re.IGNORECASE).group(0)
            data['test_id'] = int(re.search("#[0-9]+", line, flags=re.IGNORECASE).group(0)[1:])
            data['time'] = float(re.search("[0-9]+\.[0-9]+", line, flags=re.IGNORECASE).group(0))
            results.append(data)
        except Exception as e:
            pass
    
    return pd.DataFrame(results)

In [71]:
def parse_make(filename):
    with open(filename) as f:
        content = f.readlines()
        
    content = [x.strip() for x in content]
    results = []
    # Set initial data and target to none since the re.match may not find anything
    data = {'Target':None}
    data_list = []
    target = None
    
    for line in content:
        try:
            # If nothing is found, the .group(0) call will cause an error and
            # move to the exception clause
            target = re.match("Scanning dependencies of target [A-z or 0-9 or _]+|[A-z or 0-9 or _]+.so", line,
                              flags=re.IGNORECASE).group(0)[32:]
            # If there is a new target
            if data['Target'] != target:
                # And the current target is not None
                if data['Target']:
                    # Add the current data to the result set
                    results.append(data)
                # No matter what the target is clear the data_list and update the next data point with
                # the new target, empty data_list, and set built to true
                data_list = []
                data = {'Target':target, 'data_list':data_list, 'built':True}
                
        except Exception as e:
            # If there is a dependency we are building
            if target:
                # Check to see if the line has some kind of error or warning
                error_line = re.search("fatal error:|error:|warning:", line, flags=re.IGNORECASE)
                # If there is an error it "probably" failed to build (Hasn't been check with all
                # of the various errors we can get)
                if error_line:
                    data['built'] = False
                data_list.append(line)
            pass
    results.append(data)
    return pd.DataFrame(results)

In [74]:
unittest_logs = parse_logs('unittest.log')
apptest_logs = parse_logs('apptest.log')
make_log = parse_make('make2.log')
make_log

Target  built                                          data_list
0  isis3  False  [[ 10%] Building CXX object src/CMakeFiles/isi...

In [75]:
make_log = parse_make('make1.log')
make_log

Target built  \
0                    isis3  True   
1            AdaptiveGruen  True   
2        AerialPhotoCamera  True   
3                   Albedo  True   
4                AlbedoAtm  True   
5             Anisotropic1  True   
6             Anisotropic2  True   
7                   apollo  True   
8       ApolloMetricCamera  True   
9    ApolloPanoramicCamera  True   
10                     mro  True   
11               CTXCamera  True   
12    Chandrayaan1M3Camera  True   
13             CrismCamera  True   
14            DawnFcCamera  True   
15           DawnVirCamera  True   
16         Equirectangular  True   
17        ForstnerOperator  True   
18        GradientOperator  True   
19       GridPolygonSeeder  True   
20                   Hapke  True   
21               HapkeAtm1  True   
22               HapkeAtm2  True   
23     HayabusaAmicaCamera  True   
24      HayabusaNirsCamera  True   
25             HiresCamera  True   
26            HiriseCamera  True   
27              HrscCamera  True   
28             IdealCamera  True   
29              Isotropic1  True   
..                     ...   ...   
768         table2cube_app  True   
769          tabledump_app  True   
770           thm2isis_app  True   
771        thmbasemap1_app  True   
772        thmdriftcor_app  True   
773          thmnoseam_app  True   
774            thmproc_app  True   
775         thmvisflat_app  True   
776         thmvistrim_app  True   
777          tonematch_app  True   
778          translate_app  True   
779               trim_app  True   
780         trimfilter_app  True   
781             uncrop_app  True   
782             vdcomp_app  True   
783         vicar2isis_app  True   
784           vik2isis_app  True   
785             vikcal_app  True   
786           vikclean_app  True   
787          vikfixtrx_app  True   
788        viknobutter_app  True   
789        viknonoodle_app  True   
790        viknopepper_app  True   
791          viknosalt_app  True   
792          vims2isis_app  True   
793            vimscal_app  True   
794           voy2isis_app  True   
795             voycal_app  True   
796            voyramp_app  True   
797               warp_app  True   

                                             data_list  
0    [[ 10%] Building CXX object src/CMakeFiles/isi...  
1    [[ 41%] Building CXX object src/CMakeFiles/Ada...  
2    [[ 41%] Building CXX object src/CMakeFiles/Aer...  
3    [[ 41%] Building CXX object src/CMakeFiles/Alb...  
4    [[ 41%] Building CXX object src/CMakeFiles/Alb...  
5    [[ 41%] Building CXX object src/CMakeFiles/Ani...  
6    [[ 41%] Building CXX object src/CMakeFiles/Ani...  
7    [[ 41%] Building CXX object src/CMakeFiles/apo...  
8    [[ 41%] Building CXX object src/CMakeFiles/Apo...  
9    [[ 41%] Building CXX object src/CMakeFiles/Apo...  
10   [[ 41%] Building CXX object src/CMakeFiles/mro...  
11   [[ 41%] Building CXX object src/CMakeFiles/CTX...  
12   [[ 41%] Building CXX object src/CMakeFiles/Cha...  
13   [[ 41%] Building CXX object src/CMakeFiles/Cri...  
14   [[ 41%] Building CXX object src/CMakeFiles/Daw...  
15   [[ 41%] Building CXX object src/CMakeFiles/Daw...  
16   [[ 42%] Building CXX object src/CMakeFiles/Equ...  
17   [[ 42%] Building CXX object src/CMakeFiles/For...  
18   [[ 42%] Building CXX object src/CMakeFiles/Gra...  
19   [[ 42%] Building CXX object src/CMakeFiles/Gri...  
20   [[ 42%] Building CXX object src/CMakeFiles/Hap...  
21   [[ 42%] Building CXX object src/CMakeFiles/Hap...  
22   [[ 42%] Building CXX object src/CMakeFiles/Hap...  
23   [[ 42%] Building CXX object src/CMakeFiles/Hay...  
24   [[ 42%] Building CXX object src/CMakeFiles/Hay...  
25   [[ 42%] Building CXX object src/CMakeFiles/Hir...  
26   [[ 42%] Building CXX object src/CMakeFiles/Hir...  
27   [[ 42%] Building CXX object src/CMakeFiles/Hrs...  
28   [[ 42%] Building CXX object src/CMakeFiles/Ide...  
29   [[ 42%] Building CXX object src/CMakeFiles/Iso...  
..                  

In [76]:
unittest_logs[~unittest_logs.passed]

name passed  test_id  time
280         isis3_unit_test_IException  False      281  0.30
301           isis3_unit_test_SqlQuery  False      302  1.30
302    isis3_unit_test_DatabaseFactory  False      303  0.30
303           isis3_unit_test_Database  False      304  1.30
305          isis3_unit_test_SqlRecord  False      306  1.28
327  isis3_unit_test_MosaicSceneWidget  False      328  1.36

In [270]:
apptest_logs[~apptest_logs.passed]

name  passed  test_id  time
223  fits_app_test_infoAll   False      554  0.32

In [278]:
print("TOTAL TIME = {}".format(unittest_logs['time'].sum()))
print("AVERAGE TIME = {}".format(unittest_logs['time'].mean()))

TOTAL TIME = 288.92999999999995
AVERAGE TIME = 0.8755454545454544


In [279]:
failed = results[~results.passed]
failed

name  passed  test_id  time
280         isis3_unit_test_IException   False      281  0.30
301           isis3_unit_test_SqlQuery   False      302  1.30
302    isis3_unit_test_DatabaseFactory   False      303  0.30
303           isis3_unit_test_Database   False      304  1.30
305          isis3_unit_test_SqlRecord   False      306  1.28
327  isis3_unit_test_MosaicSceneWidget   False      328  1.36

In [295]:
msg = MIMEMultipart("TEST ATTACH")
msg['Subject'] = 'EMAIL TEST'
msg['From'] = 'krodriguez@usgs.gov'
msg['To'] = 'krodriguez@usgs.gov'

unittest_logs['type'] = 'unit'
apptest_logs['type'] = 'app'

df = unittest_logs[~unittest_logs.passed].append(apptest_logs[~apptest_logs.passed])
df = df.set_index(['type', 'test_id'])

html = df.to_html().replace("\n", "").replace('<th></th>', '').replace('</tr>    <tr>', '')
attachment = MIMEText(html, 'html')
msg.attach(attachment)

In [296]:
server = smtplib.SMTP('smtp.gmail.com',587) #port 465 or 587
server.ehlo()
server.starttls()
server.ehlo()
server.login('kelvinrodriguez1105@gmail.com','Axel1105')
server.sendmail(msg['From'], [msg['From']], msg.as_string())
server.quit()

(221, b'2.0.0 closing connection b12sm2961152qkc.56 - gsmtp')

In [306]:
df.to_html().replace("\n", "")

'<table border="1" class="dataframe">  <thead>    <tr style="text-align: right;">      <th></th>      <th></th>      <th>name</th>      <th>passed</th>      <th>time</th>    </tr>    <tr>      <th>type</th>      <th>test_id</th>      <th></th>      <th></th>      <th></th>    </tr>  </thead>  <tbody>    <tr>      <th rowspan="6" valign="top">unit</th>      <th>281</th>      <td>isis3_unit_test_IException</td>      <td>False</td>      <td>0.30</td>    </tr>    <tr>      <th>302</th>      <td>isis3_unit_test_SqlQuery</td>      <td>False</td>      <td>1.30</td>    </tr>    <tr>      <th>303</th>      <td>isis3_unit_test_DatabaseFactory</td>      <td>False</td>      <td>0.30</td>    </tr>    <tr>      <th>304</th>      <td>isis3_unit_test_Database</td>      <td>False</td>      <td>1.30</td>    </tr>    <tr>      <th>306</th>      <td>isis3_unit_test_SqlRecord</td>      <td>False</td>      <td>1.28</td>    </tr>    <tr>      <th>328</th>      <td>isis3_unit_test_MosaicSceneWidget</td>      

<table border="1" class="dataframe">  <thead>    <tr style="text-align: right;">      <th></th>      <th></th>      <th>name</th>      <th>passed</th>      <th>time</th>    </tr>    <tr>      <th>type</th>      <th>test_id</th>      <th></th>      <th></th>      <th></th>    </tr>  </thead>  <tbody>    <tr>      <th rowspan="6" valign="top">unit</th>      <th>281</th>      <td>isis3_unit_test_IException</td>      <td>False</td>      <td>0.30</td>    </tr>    <tr>      <th>302</th>      <td>isis3_unit_test_SqlQuery</td>      <td>False</td>      <td>1.30</td>    </tr>    <tr>      <th>303</th>      <td>isis3_unit_test_DatabaseFactory</td>      <td>False</td>      <td>0.30</td>    </tr>    <tr>      <th>304</th>      <td>isis3_unit_test_Database</td>      <td>False</td>      <td>1.30</td>    </tr>    <tr>      <th>306</th>      <td>isis3_unit_test_SqlRecord</td>      <td>False</td>      <td>1.28</td>    </tr>    <tr>      <th>328</th>      <td>isis3_unit_test_MosaicSceneWidget</td>      <td>False</td>      <td>1.36</td>    </tr>    <tr>      <th>app</th>      <th>554</th>      <td>fits_app_test_infoAll</td>      <td>False</td>      <td>0.32</td>    </tr>  </tbody></table>

In [297]:
html

'<table border="1" class="dataframe">  <thead>    <tr style="text-align: right;">                  <th>name</th>      <th>passed</th>      <th>time</th>          <th>type</th>      <th>test_id</th>                      </tr>  </thead>  <tbody>    <tr>      <th rowspan="6" valign="top">unit</th>      <th>281</th>      <td>isis3_unit_test_IException</td>      <td>False</td>      <td>0.30</td>          <th>302</th>      <td>isis3_unit_test_SqlQuery</td>      <td>False</td>      <td>1.30</td>          <th>303</th>      <td>isis3_unit_test_DatabaseFactory</td>      <td>False</td>      <td>0.30</td>          <th>304</th>      <td>isis3_unit_test_Database</td>      <td>False</td>      <td>1.30</td>          <th>306</th>      <td>isis3_unit_test_SqlRecord</td>      <td>False</td>      <td>1.28</td>          <th>328</th>      <td>isis3_unit_test_MosaicSceneWidget</td>      <td>False</td>      <td>1.36</td>          <th>app</th>      <th>554</th>      <td>fits_app_test_infoAll</td>      <td>Fals